hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/NamenodeHeartbeatService.java

```diff
@@ -447,8 +462,13 @@ private void updateJMXParameters(
       String address, NamenodeStatusReport report) {
     try {
       // TODO part of this should be moved to its own utility
-      getFsNamesystemMetrics(address, report);
-      getNamenodeInfoMetrics(address, report);
+      if (shouldUpdateJmx()) {
+        lastJmxUpdateAttempt = Time.monotonicNow();
+        getFsNamesystemMetrics(address);
+        getNamenodeInfoMetrics(address);
+      }
+      populateFsNamesystemMetrics(fsNamesystemMetrics, report);
+      populateNamenodeInfoMetrics(namenodeInfoMetrics, report);
```

Comment:
What happens if we do not populate the metrics in the report? I'm wondering whether the `shouldUpdateJmx()` condition can be outside of and wrapping about the whole `updateJmxParameters` function?

Reply:
If we don't populate the values, they will be reported as 0, which is why we should use the last report instead.

Reply:
I see, so we want the last value to be seen instead of a zero.

Reply:
My other question was "How does the AvailableSpaceResolver and related code behave when we fully disable updating JMX parameters?". In this case the reported value will be 0.

Reply:
Sorry I missed that. If the metrics are disabled, the [AvailableSpaceResolver](https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/resolver/order/AvailableSpaceResolver.java#L154) won't use "space available" to determine the Naemenode to write to, and will choose one randomly, which can lead to some Namenodes being over-used.
For the Router dashboard, the disabled metrics will show up as  0.

Reply:
Cool, thanks for explainingg.